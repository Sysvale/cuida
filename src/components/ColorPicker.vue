<template>
	<div v-if="inline">
		<div class="color-picker__label">
			{{ label }}
		</div>

		<div
			v-for="(swatchline, i) in swatch"
			:key="i"
			class="color-picker__container"
		>
			<div
				v-for="(color, j) in swatchline"
				:key="j"
			>
				<div
					:class="`color-picker__swatch--${color.replace('$', '')}`"
					@click="SwatchSelection(color)"
				>
					<CdsIcon
						v-if="isCurrentColorSelected(color)"
						height="22"
						width="22"
						name="check-outline"
						class="swatch__icon"
						:class="iconColorResolver"
					/>
				</div>
			</div>
		</div>
	</div>

	<div v-else>
		<div
			id="color-picker"
			class="color-picker__trigger"
			@click.stop="showPopover = !showPopover"
		>
			<div class="color-picker__preview" /> 
		</div>

		<CdsPopover
			v-model="showPopover"
			:right-aligned="false"
			target-id="color-picker"
			:height="popoverHeight"
			fit-content-width
		>
			<div
				v-for="(swatchline, i) in swatch"
				:key="i"
				class="popover__container"
			>
				<div
					v-for="(color, j) in swatchline"
					:key="j"
				>
					<div
						:class="`color-picker__swatch--${color.replace('$', '')}`"
						@click="SwatchSelection(color)"
					>
						<CdsIcon
							v-if="isCurrentColorSelected(color)"
							height="22"
							width="22"
							name="check-outline"
							class="swatch__icon"
							:class="iconColorResolver"
						/>
					</div>
				</div>
			</div>
		</CdsPopover>
	</div>
</template>

<script>
import CdsPopover from './Popover.vue';
import CdsIcon from './Icon.vue';
import sassColorVariables from '../assets/sass/tokens/colors.module.scss';
import ContrastChecker from '../utils/methods/contrastChecker';
import paleteBuilder from '../utils/methods/paleteBuilder.js';

export default {
	name: 'CdsColorPicker',
	components: {
		CdsPopover,
		CdsIcon,
	},

	props: {
		/**
		* Guarda a cor selecionada no Color Picker.
		*/
		modelValue: {
			type: String,
			required: true,
		},
		/**
		 * Especifica a label do ColorPicker.
		 */
		label: {
			type: String,
			default: 'Label',
		},
		/**
		* Quando true passa a mostrar as opções de cores fora do popover.
		*/
		inline: {
			type: Boolean,
			default: false,
		},
		/**
		* Especifica a altura do popover. ⚠️ Quando a prop inline estiver ativa essa prop não surte efeito.
		*/
		popoverHeight: {
			type: Number,
			default: 190,
		},
		/**
		* Conjunto de cores a serem renderizadas no ColorPicker. A prop espera um Array de Array de tokens de cor. Ex.: tokens.$gp-400
		*/
		swatch: {
			type: Array,
			default: () => ([]),
		},
	},

	data() {
		return {
			selectedColor: this.modelValue,
			sassColorVariables,
			showPopover: false,
		};
	},

	computed: {
		iconColorResolver() {
			return this.ContrastChecker(this.selectedColor, '#FFFFFF', 'POOR') 
				? 'swatch__icon--white'
				: 'swatch__icon--black';
		},

		palete() {
			return this.paleteBuilder(this.sassColorVariables.palete);
		},
	},

	watch: {
		modelValue: {
			handler(newValue, oldValue) {
				if (newValue !== oldValue) {
					this.selectedColor = newValue;
				}
			},
			immediate: true,
		},
	},

	methods:{
		paleteBuilder,
		ContrastChecker,

		SwatchSelection(color) {
			let selectedVariant = '';

			this.selectedColor = this.sassColorVariables[color.replace(/-|\$/gi, '')];
			this.palete.forEach(paleteColor => {
				if (paleteColor.colorTokens.find(token => token === color)) {
					selectedVariant = paleteColor.variantName.toLowerCase();
				}
			});

			/**
			 * **Implementa v-model**. Evento utilizado para emitir a *cor* selecionada. A cor é emitida como uma string no formato HEX.
			 * @event update:modelValue
			 * @type {Event}
			*/
			this.$emit('update:modelValue', this.selectedColor);
			/**
			 * Evento utilizado para emitir a *variante* da cor selecionada. A variante é emitida como uma string. ⚠️ Importante: a variante emitida só irá mudar se a cor mudar. A mudança de shade dentro da mesma cor não altera a variante. Ex.: tokens.$gp-400 e tokens.$gp-600 emitirão a mesta variante, a saber, "green".
			 * @event update:colorVariant
			 * @type {Event}
			*/
			this.$emit('variant-chosen', selectedVariant);
		},

		isCurrentColorSelected(color) {
			return this.selectedColor === this.sassColorVariables[color.replace(/-|\$/gi, '')];
		},
	},
}
</script>

<style lang="scss" scoped>
@use '../assets/sass/tokens/index' as tokens;

.color-picker {
	&__container {
		display: flex;
	}

	&__label {
		margin: tokens.mb(1);
		@include tokens.label;
	}

	&__swatch {
		@include tokens.paleteResolver using ($color) {
			background-color: $color;
			height: 24px;
			width: 24px;
			border-radius: 4px;
			margin: tokens.mTRBL(1, 1, 0, 0);
			padding-top: 2px;
			padding-left: 1px;
			cursor: pointer;
			transition: tokens.$interaction;

			&:hover {
				transform: scale(1.1);
				transition: tokens.$interaction;
			}
		}
	}

	&__trigger {
		padding: tokens.pa(1);
		outline: 1px solid tokens.$n-50;
		width: fit-content;
		border-radius: tokens.$border-radius-lil;
		transition: tokens.$hover;
		cursor: pointer;

		&:hover {
			outline: 1px solid tokens.$n-200;
			transition: tokens.$hover;
		}
	}

	&__preview {
		height: 24px;
		width: 24px;
		border-radius: 4px;
		background-color: v-bind(selectedColor);
	}
}

.swatch {
	&__icon {
		&--white {
			color: tokens.$n-0;
		}

		&--black {
			color: tokens.$n-900;
		}
	}
}

.popover__container {
	display: flex;
}
</style>
