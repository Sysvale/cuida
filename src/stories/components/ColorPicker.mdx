import { Controls, Canvas, Meta, Story } from '@storybook/blocks';
import ColorPicker from '../../components/ColorPicker.vue';
import * as ColorPickerStories from './ColorPicker.stories';

<Meta of={ColorPickerStories} />

export const Template = (args) => ({
	components: { CdsColorPicker: ColorPicker },
	setup() {
		return { args };
	},
	data() {
		return {
			color: '#2AC092',
		};
	},
	template: /*html*/ `
		<div class="d-flex" style="width: 550px; justify-content: space-between; align-items: end;">
			<div>
				<cds-color-picker
					v-bind="args"
					v-model="color"
					@change="handleEmit"
					@variant-chosen="handleVariantChosen"
				/>
			</div>
			<div>
				<div style="height: 52px; width: 52px; border-radius: 8px" :style="{background: color}"/>
				<i style="margin-top: 1px; font-size: 12.5px; color: #3b4754">⚠️ Div usada como demo. Não faz parte do ColorPicker.</i>
			</div>
		</div>`,
	methods: {
		handleEmit(value) { console.info('value: ', value); },
		handleVariantChosen(value) { console.info('%cv-model:colorVariant ', 'color: #2C70CD;', value); }
	},
	updated() {
		console.info('%cv-model: ', 'color: #2C70CD;', this.color);
	},
});

# ColorPicker

### ColorPickers permitem aos usuários escolherem cores de modo visual.

***

<br />

## Quando usar:

* Use o ColorPicker em cenários em que o usuário precisa escolher uma cor dentre um conjunto de cores disponíveis.
* Use o ColorPicker em conjunto com inputs caso queira que o usuário visualize o valor HEX da cor escolhida.
* Em caso de poucas opções de cor dê preferência ao ColorPicker no modo `inline`.

<br />

## Obs:

* O ColorPicker não permite ao usuário escolher qualquer cor, apenas cores disponíveis no Cuida.
* O modo como você estrutura o array (matriz) da prop swatch importa na renderização do componente. Ex.: se você tiver 3 linhas na matriz o ColorPicker renderizará 3 linhas de cores.
* O ColorPicker emite a cor selecionada como uma string contendo o HEX da cor.

<br />

## Preview

export const swatch2 = [
	['$gp-200', '$gp-300', '$gp-400', '$gp-500'],
	['$ts-200', '$ts-300', '$ts-400', '$ts-500'],
	['$bn-200', '$bn-300', '$bn-400', '$bn-500'],
	['$in-200', '$in-300', '$in-400', '$in-500'],
	['$vr-200', '$vr-300', '$vr-400', '$vr-500'],
	['$pp-200', '$pp-300', '$pp-400', '$pp-500'],
	['$rc-200', '$rc-300', '$rc-400', '$rc-500'],
	['$og-200', '$og-300', '$og-400', '$og-500'],
	['$al-200', '$al-300', '$al-400', '$al-500'],
	['$n-10', '$n-40', '$n-300', '$n-600'],
];

export const swatch = [
	[
		'$gp-400',
		'$ts-400',
		'$bn-400',
		'$in-400',
		'$vr-400',
	],
	[
		'$pp-400',
		'$rc-400',
		'$og-400',
		'$al-400',
		'$n-400',
	],
];

### **Inline**

<Canvas>
  <Story of={ColorPickerStories.ColorPicker} />
</Canvas>

<Controls />

### **Com popover**

<Canvas>
  <Story of={ColorPickerStories.ColorPicker2} />
</Canvas>

<Controls />
