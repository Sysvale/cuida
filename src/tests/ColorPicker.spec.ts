// @vitest-environment jsdom
import { describe, test, expect } from 'vitest';
import ColorPicker from '../components/ColorPicker.vue';
import { mount } from '@vue/test-utils';

describe('ColorPicker', () => {
	test('renders correctly', async () => {
		const wrapper = mount(ColorPicker, {
			props: {
				modelValue: '#000000',
				inline: true,
				swatch: ['#000000', '#000000'],
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
