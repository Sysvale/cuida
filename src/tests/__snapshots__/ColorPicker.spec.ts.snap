// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorPicker > renders correctly 1`] = `
"<div data-v-2c841161="">
  <div data-v-2c841161="" class="color-picker__label">Label</div>
  <div data-v-2c841161="" class="color-picker__container">
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--#">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
  </div>
  <div data-v-2c841161="" class="color-picker__container">
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--#">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
    <div data-v-2c841161="">
      <div data-v-2c841161="" class="color-picker__swatch--0">
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;
